src/Ldap/Client.hs

```diff
@@ -139,51 +149,56 @@ data Disconnect = Disconnect Type.ResultCode Dn Text
 
 instance Exception Disconnect
 
+open :: Host -> PortNumber -> IO Ldap
+open host port = do
+  context <- Conn.initConnectionContext
+  conn <- Conn.connectTo context (params host port)
+  l <- newLdap conn
+  return l
+
+close :: Ldap -> IO ()
+close l@(Ldap client v conn threads) = do
+  unbindAsync l
+  forM threads Async.cancel
+  Conn.connectionClose conn
+
 -- | The entrypoint into LDAP.
 --
 -- It catches all LDAP-related exceptions.
 with :: Host -> PortNumber -> (Ldap -> IO a) -> IO (Either LdapError a)
 with host port f = do
-  context <- Conn.initConnectionContext
-  bracket (Conn.connectTo context params) Conn.connectionClose (\conn ->
-    bracket newLdap unbindAsync (\l -> do
-      inq  <- newTQueueIO
-      outq <- newTQueueIO
-      as   <- traverse Async.async
-        [ input inq conn
-        , output outq conn
-        , dispatch l inq outq
-        , f l
-        ]
-      fmap (Right . snd) (Async.waitAnyCancel as)))
+  bracket (open host port) close (\l -> do
+    as   <- Async.async $ f l
+    fmap Right (Async.wait as))
```

Comment:
Hrmm, I'm afraid this isn't quite right. Should an exception be raised by a helper thread, we want it to propagate to the main thread. The ignored exception will result in weird behavior (probably a `BlockedIndefinitelyOnMVar` exception will be raised eventually) that would be annoying to debug. Currently, we rely on `waitAnyCancel` to do the propagation correctly.

(One way to convince the type system that putting `threads` and `as` in a single list is still OK is to store `[Async Void]` in `threads` so that `waitAnyCancel (as : map absurd threads)` typechecks.)